### PR TITLE
Fix incorrect __torch_function__ handling in einsum

### LINKED ISSUE #38479: torch.einsum does not pass equation argument to __torch_function__ API
## 🐛 Bug

when delegating torch.einsum call to an object which implements
`__torch_function__` API the equation argument is not passed resulting in the error.
```TypeError: einsum(): argument 'equation' (position 1) must be str, not Tensor```

this was tested on pytorch 1.5.0

I've actually found the cause of this bug and have written a fix.

the following script illustrates the problem and the proposed solution

## To Reproduce

```python 
import torch

class Wrapper():
    def __init__(self,data):
        self.data = data
    
    def __torch_function__(self, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}

        #unwrap inputs if necessary
        def unwrap(v):
            return v.data if isinstance(v,Wrapper) else v
        args = map(unwrap,args)
        kwargs = {k:unwrap(v) for k,v in kwargs.items()}

        return func(*args, **kwargs)



# fixed einsum implementation
from torch import Tensor,_VF
from torch._overrides import has_torch_function,handle_torch_function
def fixed_einsum(equation,*operands):
    if not torch.jit.is_scripting():
        if any(type(t) is not Tensor for t in operands) and has_torch_function(operands):
            # equation is not passed
            # return handle_torch_function(einsum, operands, *operands)
            return handle_torch_function(fixed_einsum, operands, equation,*operands)
    if len(operands) == 1 and isinstance(operands[0], (list, tuple)):
        # the old interface of passing the operands as one list argument
        operands = operands[0]

        # recurse incase operands contains value that has torch function
        #in the original implementation this line is omitted
        return fixed_einsum(equation,*operands)

    return _VF.einsum(equation, operands)


if __name__ == "__main__":
    print(torch.__version__)
    # uncomment to use fixed einsum
    # torch.einsum = fixed_einsum

    #operands are wrapped
    x = Wrapper(torch.randn(5))
    y = Wrapper(torch.randn(4))
    assert torch.allclose(torch.einsum('i,j->ij',x, y),torch.ger(x,y))  # outer product
    print("works with wrapped inputs")    

    #old interface operands is a list
    a = Wrapper(torch.randn(2,3))
    b = Wrapper(torch.randn(5,3,7))
    c = Wrapper(torch.randn(2,7))
    assert torch.allclose(torch.einsum('ik,jkl,il->ij', [a, b, c]),torch.nn.functional.bilinear(a,c,b)) # bilinear interpolation
    print("works with old API operands is list")
    
    #equation is wrapped
    As = Wrapper(torch.randn(3,2,5))
    Bs = Wrapper(torch.randn(3,5,4))
    equation = Wrapper('bij,bjk->bik')
    assert torch.allclose(torch.einsum(equation, As, Bs),torch.matmul(As,Bs)) # batch matrix multiplication
    print("works with equation wrapped")

    #see that it also works with plain tensors
    x = torch.randn(5)
    y = torch.randn(4)
    assert torch.allclose(torch.einsum('i,j->ij',x, y),torch.ger(x,y)) 
    print("works with no wrapped values")



```

cc @albanD @mruberry

### PR DESCRIPTION
**Pull request by rgommers**
Closes gh-38479

### CODE REVIEW

File: test/test_overrides.py [hunk 1]
```diff
@@ -488,5 +488,45 @@ def test(self):
...
+        a = Wrapper(torch.randn(2, 3))
+        b = Wrapper(torch.randn(5, 3, 7))
+        c = Wrapper(torch.randn(2, 7))
+        self.assertTrue(torch.allclose(torch.einsum('ik,jkl,il->ij', [a, b, c]),
+                                       torch.nn.functional.bilinear(a, c, b)))
+
+        # `equation` input is wrapped
+        As = Wrapper(torch.randn(3, 2, 5))
+        Bs = Wrapper(torch.randn(3, 5, 4))
+        equation = Wrapper('bij,bjk->bik')
```
Comment (hameerabbasi) [inline test/test_overrides.py:527]: I don’t think non-tensor-likes should be wrapped. This gives the impression that they will be dispatched on, which they won’t.

File: torch/functional.py [hunk 2]
```diff
@@ -294,10 +294,15 @@ def einsum(equation, *operands):
...
         if any(type(t) is not Tensor for t in operands) and has_torch_function(operands):
-            return handle_torch_function(einsum, operands, *operands)
+            return handle_torch_function(einsum, operands, equation, *operands)
+
     if len(operands) == 1 and isinstance(operands[0], (list, tuple)):
         # the old interface of passing the operands as one list argument
         operands = operands[0]
+        # recurse incase operands contains value that has torch function
+        # in the original implementation this line is omitted
+        return einsum(equation, *operands)
```
Comment (hameerabbasi) [inline torch/functional.py:304]: This shouldn’t be necessary IIUC. `operands[0]` will already Dispatch above.

**Review by hameerabbasi — COMMENTED** [review]
A couple of small comments.

File: torch/functional.py (hunk 2 above)
Comment (rgommers) [reply torch/functional.py:304]: No that's not the case. See the `if any(type(t) ....` line that guards `handle_torch_function`. Dispatching happens on a sequence of `Tensor`s or anything with a `__torch_function__` method, not with a sequence of "objects with `__torch_function__` method".

Without this line one gets:
```
  File "/home/rgommers/code/pytorch/torch/functional.py", line 306, in einsum
    return _VF.einsum(equation, operands)
TypeError: expected Tensor as element 0 in argument 1, but got Wrapper
```

File: test/test_overrides.py (hunk 1 above)
Comment (rgommers) [reply test/test_overrides.py:527]: Fair enough, I'll remove that test case.

File: torch/functional.py (hunk 2 above)
Comment (hameerabbasi) [reply torch/functional.py:304]: Ah, understood.

**Comment by rgommers** [comment]
All CI failures after first commit were unrelated, second commit just removes a test. I think this is good to go.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**